arangod/Aql/OptimizerRulesCluster.cpp

```diff
@@ -0,0 +1,464 @@
+  auto parents = in->getParents();
+  if(parents.size() == 1){
+    return parents.front();
+  }
+  return nullptr;
+}
+
+ExecutionNode* hasSingleParent(ExecutionNode* in, EN::NodeType type){
+  auto* parent = hasSingleParent(in);
+  if(parent) {
+    if(parent->getType() == type){
+      return parent;
+    }
+  }
+  return nullptr;
+}
+
+ExecutionNode* hasSingleParent(ExecutionNode* in, std::vector<EN::NodeType> types){
+  auto* parent = hasSingleParent(in);
+  if(parent) {
+    for(auto const& type : types){
+      if(parent->getType() == type){
+        return parent;
+      }
+    }
+  }
+  return nullptr;
+}
+
+// in plan above
+ExecutionNode* hasSingleDep(ExecutionNode* in){
+  auto deps = in->getDependencies();
+  if(deps.size() == 1){
+    return deps.front();
+  }
+  return nullptr;
+}
+
+ExecutionNode* hasSingleDep(ExecutionNode* in, EN::NodeType type){
+  auto* dep = hasSingleDep(in);
+  if(dep) {
+    if(dep->getType() == type){
+      return dep;
+    }
+  }
+  return nullptr;
+}
+
+ExecutionNode* hasSingleDep(ExecutionNode* in, std::vector<EN::NodeType> types){
+  auto* dep = hasSingleDep(in);
+  if(dep) {
+    for(auto const& type : types){
+      if(dep->getType() == type){
+        return dep;
+      }
+    }
+  }
+  return nullptr;
+}
+
+Index* hasSingleIndexHandle(ExecutionNode* node){
+  TRI_ASSERT(node->getType() == EN::INDEX);
+  IndexNode* indexNode = static_cast<IndexNode*>(node);
+  auto indexHandleVec = indexNode->getIndexes();
+  if (indexHandleVec.size() == 1){
+    return indexHandleVec.front().getIndex().get();
+  }
+  return nullptr;
+}
+
+Index* hasSingleIndexHandle(ExecutionNode* node, Index::IndexType type){
+  auto* idx = hasSingleIndexHandle(node);
+  if (idx && idx->type() == type ){
+    return idx;
+  }
+  return nullptr;
+}
+
+std::vector<AstNode const*> hasBinaryCompare(ExecutionNode* node){
+  // returns any AstNode in the expression that is
+  // a binary comparison.
+  TRI_ASSERT(node->getType() == EN::INDEX);
+  IndexNode* indexNode = static_cast<IndexNode*>(node);
+  AstNode const* cond = indexNode->condition()->root();
+  std::vector<AstNode const*> result;
+
+  auto preVisitor = [&result] (AstNode const* node) {
+    if (node == nullptr) {
+      return false;
+    };
+
+    if(node->type == NODE_TYPE_OPERATOR_BINARY_EQ){
+      result.push_back(node);
+      return false;
+    }
+
+    //skip over NARY AND/OR
+    if(node->type == NODE_TYPE_OPERATOR_NARY_OR ||
+       node->type == NODE_TYPE_OPERATOR_NARY_AND) {
+      return true;
+    } else {
+      return false;
+    }
+
+  };
+  auto postVisitor = [](AstNode const*){};
+  Ast::traverseReadOnly(cond, preVisitor, postVisitor);
+
+  return result;
+}
+
+std::string getFirstKey(std::vector<AstNode const*> compares){
+  for(auto const* node : compares){
+    AstNode const* keyNode = node->getMemberUnchecked(0);
+    if(keyNode->type == AstNodeType::NODE_TYPE_ATTRIBUTE_ACCESS && keyNode->stringEquals("_key")) {
```

Comment:
* what if the condition is not `doc._key == "key"` but `"key" == doc._key`? will it still work?
* instead of using `"_key"`, please use `StaticStrings::KeyString`